### PR TITLE
The Cycling 2.0, Airlocks can now be built in game to cycle and named

### PR DESCRIPTION
**Pull request by AlinaStarkova**
## About The Pull Request

- You can now edit an airlock control board (both physically and through an RCD) to set the airlock name and what ID it cycles on
- Airlocks now say what ID it cycles on when examined
- when an airlock is made it takes its circuits set name and ID and updates the cycling
- i redid how the cycle_id is updated on creation, so now when an airlock is made and grabbing airlocks with an identical ID to cycle it checks if the other airlocks with that ID needs a reference to the new airlock, and if so, it adds it, so you can add doors back to a cycle system if it breaks

NOTE: unfortunately the way linear airlocks, like the small 2 door airlocks to space, set their links, you cannot repair their cycle this way, you will have to deconstruct the remaining door and rebuild both to cycle again

Setting the settings:
hi1
hi2
hi3
hi4

Examine:
hi5
hi6

Clip of Cycling:

https://user-images.githubusercontent.com/40489693/132058137-ecd7568d-a732-490a-b6e2-1f4a50708743.mp4

## Changelog
:cl:
expansion: You can now set an Airlocks name and cycle ID in the RCD airlock settings menu, or by clicking on an airlock control board and making an airlock manually, this allows you to make airlocks that can cycle with each other. examine an airlock to see what, if any, ID they are set to cycle on (linear 1x3 airlocks to space currently do not use cycleIDs)
/:cl:

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
How many airlocks can you link together?

**Comment by AlinaStarkova** [comment]
> 
> 
> How many airlocks can you link together?

idk, how big can a list of object references be? lol

theres no real way to easily limit it since your selecting the ID before its even made, so what happens if it hits its limit? gotta go back and null everything?

i dont really see an issue with it not having a limit tho, since you cant actually do anything with cycling, all it does is close other airlocks if you walk into one, and you cant modify existing airlocks without completely rebuilding them, the absolute worst people can do is fill the halls with linked airlocks, but thats not any more grief than filling the halls with linked airlocks is anyway

and remember: clicking an airlock directly will open it without cycling anything, so the AI can open the spammed doors if it happens and bolt them open, and the memer can be beaten by sec as an IC issue

TL;DR, its not really a concern

**Review by stylemistake — APPROVED** [review]

File: code/game/machinery/doors/airlock_electronics.dm [hunk 1]
```diff
@@ -80,6 +86,12 @@
 				return
 			accesses -= SSid_access.get_region_access_list(list(region))
 			. = TRUE
+		if("passedName")
+			passed_name = params["passedName"]
```
Comment (Mothblocks) [inline code/game/machinery/doors/airlock_electronics.dm:90]: You can't trust user input like this, it could be literally anything.

Check if its text, and limit its length.

File: code/game/machinery/doors/airlock_electronics.dm [hunk 2]
```diff
@@ -80,6 +86,12 @@
...
+			. = TRUE
+		if("passedCycleId")
+			passed_cycle_id = params["passedCycleId"]
```
Comment (Mothblocks) [inline code/game/machinery/doors/airlock_electronics.dm:93]: Same here.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/game/machinery/doors/airlock_electronics.dm (hunk 1 above)
Comment (AlinaStarkova) [reply code/game/machinery/doors/airlock_electronics.dm:90]: it can be a number, it doesnt matter what it is, as long as its the same for all doors

File: code/game/machinery/doors/airlock_electronics.dm [hunk 3]
```diff
@@ -80,6 +86,14 @@
 				return
 			accesses -= SSid_access.get_region_access_list(list(region))
 			. = TRUE
+		if("passedName")
+			var/new_name = trim(params["passedName"], 30)
+			passed_name = new_name
```
Comment (Watermelon914) [inline code/game/machinery/doors/airlock_electronics.dm:91]: this isn't sanitized. This needs to be sanitized

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/game/machinery/doors/airlock_electronics.dm (hunk 3 above)
Comment (Watermelon914) [reply code/game/machinery/doors/airlock_electronics.dm:91]: Also checked the other threads from mothblocks, I'm talking specifically about html.
This lets you apply html that gets executed on clients if they examine the object, you need to html encode it.

File: code/game/machinery/doors/airlock.dm [hunk 4]
```diff
@@ -595,6 +598,10 @@
 
 /obj/machinery/door/airlock/examine(mob/user)
 	. = ..()
+	if(closeOtherId)
+		. += span_warning("This airlock cycles on ID: [closeOtherId].")
```
Comment (Watermelon914) [inline code/game/machinery/doors/airlock.dm:602]: closeOtherId is not sanitized here.
